### PR TITLE
[86bxdttwy][popper] cursor anchoring

### PR DESCRIPTION
**Pull request by msereniti**
## Motivation and Context

This PR adds feature to cover rare edge case. When popper trigger is bigger than popper, it might be better for popper to appear where user cursor enters the trigger.

## How has this been tested?

With browser tests.

## Screenshots

Screenshot 2024-04-23 at 15 19 49
propriate):

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Nice improve.

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly or it's not required.
- [x] Unit tests are not broken.
- [x] I have added changelog note to corresponding `CHANGELOG.md` file with planned publish date.
- [x] I have added new unit tests on added of fixed functionality.

### CODE REVIEW

**Review by sheila-semrush — APPROVED** [review]
Looks good!

File: semcore/popper/src/index.d.ts [hunk 1]
```diff
@@ -92,6 +92,10 @@ export type PopperProps = OutsideClickProps &
      * If enabled, you will need to use setTrigger function from children rendering function to set popper trigger.
      */
     explicitTriggerSet?: boolean;
+    /**
+     * If set, popper will be placed near the place mouse cursor entered the trigger
+     */
+    cursorAnchoring?: 'vertical' | 'horizontal';
```
Comment (ilyabrower) [inline semcore/popper/src/index.d.ts:98]: looks like it should be boolean and get type (vertical or horizontal) by placement prop...

File: semcore/popper/src/modifiers/cursorAnchoring.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import { Modifier } from '@popperjs/core';
+
+const clamp = (value: number, boundary1: number, boundary2: number) => {
+  const min = Math.min(boundary1, boundary2);
+  const max = Math.max(boundary1, boundary2);
+  return Math.min(Math.max(value, min), max);
+};
+
+type Options = {
+  cursorAnchoring?: 'horizontal' | 'vertical';
```
Comment (ilyabrower) [inline semcore/popper/src/modifiers/cursorAnchoring.ts:11]: why optional?

File: semcore/popper/src/modifiers/cursorAnchoring.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      const offsetParentRect =
+        offsetParent === window ? getWindowRect() : offsetParent.getBoundingClientRect();
+      const mouseX = mouseEnterCursorPositionRef?.current?.x;
+      if (mouseX === undefined) return;
+      const width = Math.min(rects.reference.width, rects.popper.width);
+      const x = clamp(
+        mouseX - offsetParentRect.x - width / 2,
+        rects.reference.x,
+        rects.reference.x + rects.reference.width - width,
+      );
```
Comment (ilyabrower) [inline semcore/popper/src/modifiers/cursorAnchoring.ts:35]: let's talk about this math's... maybe we should add some comments

File: semcore/popper/src/modifiers/cursorAnchoring.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  phase: 'beforeRead',
+  enabled: true,
+  fn: ({ state, options }) => {
+    const { cursorAnchoring, mouseEnterCursorPositionRef } = options;
+    const { rects, elements } = state;
+    if (cursorAnchoring === 'horizontal') {
+      const offsetParent = getOffsetParent(elements.popper);
+      const offsetParentRect =
+        offsetParent === window ? getWindowRect() : offsetParent.getBoundingClientRect();
+      const mouseX = mouseEnterCursorPositionRef?.current?.x;
```
Comment (ilyabrower) [inline semcore/popper/src/modifiers/cursorAnchoring.ts:28]: you don't need first `?`, because it is not optional by types

File: semcore/popper/src/Popper.jsx [hunk 5]
```diff
@@ -147,6 +148,7 @@ class Popper extends Component {
   }
 
   state = { ignoreTriggerFocusUntil: 0 };
+  mouseEnterCursorPositionRef = { current: null };
```
Comment (ilyabrower) [inline semcore/popper/src/Popper.jsx:151]: I'm not sure, that you need ref here... you could just use object, as I see...

File: semcore/popper/src/Popper.jsx [hunk 6]
```diff
@@ -183,6 +185,13 @@ class Popper extends Component {
       name: 'computeStyles',
       options: { gpuAcceleration: false },
     });
+    modifiersOptions.push({
+      name: 'cursorAnchoring',
+      options: {
+        cursorAnchoring: this.asProps.cursorAnchoring,
+        mouseEnterCursorPositionRef: this.mouseEnterCursorPositionRef,
+      },
+    });
```
Comment (ilyabrower) [inline semcore/popper/src/Popper.jsx:194]: maybe we should push this modifier only if we have `this.asProps.cursorAnchoring` ?

File: semcore/popper/src/Popper.jsx (hunk 5 above)
Comment (msereniti) [reply semcore/popper/src/Popper.jsx:151]: I would prefer to keep normal ref object just to make it more predictable

File: semcore/popper/src/modifiers/cursorAnchoring.ts (hunk 3 above)
Comment (msereniti) [reply semcore/popper/src/modifiers/cursorAnchoring.ts:35]: Not sure that comments will make it more clear but I can ensure you that it's correct

**Review by ilyabrower — APPROVED** [review]
